cranelift/codegen/src/isa/spirv/abi.rs

```diff
@@ -0,0 +1,249 @@
+use crate::binemit::Stackmap;
+use crate::ir::types::Type;
+use crate::ir::Function;
+use crate::ir::{ArgumentExtension, ArgumentPurpose, StackSlot};
+use crate::isa::spirv::inst::EmitState;
+use crate::isa::spirv::inst::Inst;
+use crate::machinst::abi::ABIBody;
+use crate::settings;
+use alloc::vec::Vec;
+
+use regalloc::*;
+
+use rspirv::dr::{Instruction, Operand};
+use spirv_headers::Op;
+
+use std::sync::atomic::{AtomicU32, Ordering};
+
+struct SpirvReturn {
+    id: u32,
+    real_reg: RealReg,
+}
+
+pub(crate) struct SpirvABIBody {
+    flags: settings::Flags,
+    next_id: AtomicU32,
```

Comment:
This won't work they way you may think it does. You are only accessing it from a single thread, so it is equivalent to an `u32`. What about adding a relocation type for ids, so that the code creating a SPIR-V module from all functions can make the ids unique?

Reply:
Yeah; we're having a discussion internally on how to best solve this. We'll need to hand out `id`'s in a few cases; one is to allocate types and the other as virtual SSA registers, so we need some architecture around this. 

For type allocation we're thinking of doing a two-pass approach (eg. scan which types are needed and alloc them followed by a pass that just uses them).

For handing out id's during instruction lowering, we'll need some kind of way to have relocateable id's most likely so we'll take a look at that.